### DISCUSSION TITLE
distributed-transaction/idempotent

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#177] [General]
# distributed-transaction/idempotent

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/distributed-transaction/idempotent.html

**Comment by littlebear-xbz** [comment]
ABA问题这一块读起来有点糊涂，

乐观锁逻辑 这个图没有很清晰的表示 第二次更新为90元应该成功的， 后续由于重试更新为80元是不成功的

另外最后一句话存在问题
重新获取最新的版本号就**行**修改，即可解决 高并发中 ABA 问题

**Reply by isno** [reply]
这一块我重新改改，本周能修改完。

**Reply by isno** [reply]
已经修改完，如果有兴趣可以再读一遍，看看修改的是否合适。

**Reply by lwr-max** [reply]
这个举例个人觉得有问题，有点混乱。aba提现不出来。
a应该是初始值，b应该是中间值
我的理解应该是100是初初始值，90是目标值，80是中间值。
应该是在将100变成90时，出现并发导致有两个线程同时在更新，其中第二个线程因为各种原因堵塞了，在第一个线程操作完后，数据库变成了90，这时候又来了第三个线程，将数据库中的90变成了100。第三个线程操作完后，第二个线程恢复正常，又把100改成了90。

**Comment by eastjoe** [comment]
同感ABA例子不太好，品起来还是不够直观，网上找了个例子：

1、小琳在 ATM 1 转账 100 块钱给小李；
2、由于ATM 1 出现了网络拥塞的原因卡住了，这时候小琳跑到旁边的 ATM 2 再次操作转账；
3、ATM 2 没让小琳失望，执行了 CAS(100,0)，很痛快地完成了转账，此时小琳的账户余额为 0；
4、小王这时候又给小琳账上转了 100，此时小琳账上余额为 100；
5、这时候 ATM 1 网络恢复，继续执行 CAS(100,0)，居然执行成功了，小琳账户上余额又变为了 0；
6、这时候小王微信跟小琳说转了 100 过去，是否收到呢？小琳去查了下账，摇了摇头，那么问题来了，钱去了哪呢？

**Comment by JackCharles** [comment]
“然后又其他线程把 B 值改成了 A” 更正为 “然后其他线程又把 B 值改成了 A”，更为通顺。

-- 感谢大佬提供优秀的学习资料。

**Comment by zjyyzhengli** [comment]
”已经广泛应用于需要**~分~**生成唯一标识符的场景“ -- "分"字多余了。

**Reply by isno** [reply]
谢谢

**Comment by lzfgit-byte** [comment]
ABA问题是被删了？
我的理解是一个线程在进行cas操作的时候，拿到的A不知道是最开始的A还是变为B后有变成A的A。
这个时候就需要额外的信息来记录这个变化。也就是值的版本。即使线程拿最新值到的还是A但是跟线程当前持有的A(老值)的version不一样，也就说明值变了，也需要更新当前线程的A值

**Reply by isno** [reply]
是的，删掉了。你理解的没错！
